src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/NativeScope.java

```diff
@@ -176,31 +181,167 @@ public MemoryAddress allocate(MemoryLayout layout, double value) {
+        return addr;
+    }
+
+    /**
+     * Allocate a block of memory in this native scope with given layout and initialize it with given short array.
+     * The address returned by this method is associated with a segment which cannot be closed. Moreover, the returned
+     * address must conform to the layout alignment constraints.
+     * @param elementLayout the element layout of the array to be allocated.
+     * @param array the array to be copied on the newly allocated memory block.
+     * @return an address which points to the newly allocated memory block.
+     * @throws OutOfMemoryError if there is not enough space left in this native scope, that is, if
+     * {@code limit() - size() < (elementLayout.byteSize() * array.length)}.
+     * @throws IllegalArgumentException if {@code elementLayout.byteSize()} does not conform to the size of a short value.
+     */
+    public MemoryAddress allocateArray(ValueLayout elementLayout, short[] array) {
+        if (elementLayout.byteSize() != 2) {
+            throw new IllegalArgumentException("Bad layout size");
+        }
+        MemoryAddress addr = allocate(MemoryLayout.ofSequence(array.length, elementLayout));
+        if (elementLayout.order() == ByteOrder.nativeOrder()) {
+            addr.segment().copyFrom(MemorySegment.ofArray(array));
+        } else {
+            ((AbstractMemorySegmentImpl)addr.segment()).copyFromSwap(MemorySegment.ofArray(array), elementLayout.byteSize());
+        }
+        return addr;
+    }
+
+    /**
+     * Allocate a block of memory in this native scope with given layout and initialize it with given char array.
+     * The address returned by this method is associated with a segment which cannot be closed. Moreover, the returned
+     * address must conform to the layout alignment constraints.
+     * @param elementLayout the element layout of the array to be allocated.
+     * @param array the array to be copied on the newly allocated memory block.
+     * @return an address which points to the newly allocated memory block.
+     * @throws OutOfMemoryError if there is not enough space left in this native scope, that is, if
+     * {@code limit() - size() < (elementLayout.byteSize() * array.length)}.
+     * @throws IllegalArgumentException if {@code elementLayout.byteSize()} does not conform to the size of a char value.
+     */
+    public MemoryAddress allocateArray(ValueLayout elementLayout, char[] array) {
+        if (elementLayout.byteSize() != 2) {
+            throw new IllegalArgumentException("Bad layout size");
+        }
+        MemoryAddress addr = allocate(MemoryLayout.ofSequence(array.length, elementLayout));
+        if (elementLayout.order() == ByteOrder.nativeOrder()) {
+            addr.segment().copyFrom(MemorySegment.ofArray(array));
+        } else {
+            ((AbstractMemorySegmentImpl)addr.segment()).copyFromSwap(MemorySegment.ofArray(array), elementLayout.byteSize());
+        }
```

Comment:
This code seems to be the same in the other implementations (except for the call to MS::ofArray). Maybe you want to factor it out into a helper method?